File: ts/packages/server/scripts/write_soak_snapshot.ts
/** Write a Phase 9 soak snapshot with the TypeScript build.
 *
 * Indexes every repository in a manifest with this build's `Application` and
 * records the chunk rows and the search rankings for the manifest's queries,
 * in the snapshot shape `src/soak.ts` compares. Run the Python writer first:
 * it writes the `.ci-mcp/project.toml` markers whose project ids both
 * snapshots must share for chunk ids to be comparable.
 *
 * Usage, from the repository root:
 *
 *     bun ts/packages/server/scripts/write_soak_snapshot.ts \
 *       --manifest soak.manifest.json --output soak-ts.json
 */

import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { Application, runtimePathsFromEnvironment } from "../src/application.ts";
import { dumpJson } from "../src/jsonable.ts";
import { SoakManifest, SoakSnapshot } from "../src/soak.ts";
import { indexSettingsFromEnvironment } from "../src/settings.ts";
import { checkoutHead } from "../src/update-check.ts";

interface Arguments {
  manifest: string;
  output: string;
  dataDir: string | undefined;
}

function parseArguments(argv: string[]): Arguments {
  const values = new Map<string, string>();
  for (let index = 0; index < argv.length; index += 2) {
    const flag = argv[index];
    if (flag === undefined || !flag.startsWith("--") || argv[index + 1] === undefined) {
      throw new Error(`expected --flag value pairs, got: ${argv.join(" ")}`);
    }
    values.set(flag.slice(2), argv[index + 1] as string);
  }
  const manifest = values.get("manifest");
  const output = values.get("output");
  if (manifest === undefined || output === undefined) {
    throw new Error(
      "usage: write_soak_snapshot.ts --manifest <file> --output <file> [--data-dir <dir>]",
    );
  }
  return { manifest, output, dataDir: values.get("data-dir") };
}

const repositoryRoot = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "../../../..");

const { manifest: manifestPath, output, dataDir } = parseArguments(process.argv.slice(2));
const manifest = SoakManifest.parse(JSON.parse(fs.readFileSync(manifestPath, "utf8")));
if (manifest.queries.length === 0) {
  throw new Error("the soak manifest needs at least one query");
}
const limit = manifest.limit ?? 8;
const temporary = dataDir ?? fs.mkdtempSync(path.join(os.tmpdir(), "ci-mcp-ts-soak-"));
const application = new Application(
  { data: temporary, cache: runtimePathsFromEnvironment().cache },
  {
    cwd: process.cwd(),
    settings: {
      ...indexSettingsFromEnvironment(),
      indexExecution: "in-process",
      brokerMode: "off",
    },
  },
);

try {
  const repositories = [];
  for (const repository of manifest.repositories) {
    const root = path.resolve(process.cwd(), repository.path);
    const name = repository.name ?? path.basename(root);
    const project = await application.initProject(root);
    await application.indexProject(project.id);
    const chunks = (await application.store.listChunks([project.id])).sort((left, right) =>
      left.chunk_id.localeCompare(right.chunk_id),
    );
    const queries = [];
    for (const query of manifest.queries) {
      const response = await application.searchCode(query, { projects: [project.id], limit });
      queries.push({
        query,
        hits: response.hits.map((hit) => ({ chunk_id: hit.chunk_id, score: hit.score })),
      });
    }
    process.stderr.write(`${name}: ${chunks.length} chunks, ${queries.length} queries\n`);
    repositories.push({
      name,
      path: root,
      project_id: project.id,
      chunk_count: chunks.length,
      chunks,
      queries,
    });
  }
  const snapshot = SoakSnapshot.parse({
    schema_version: 1,
    build: "typescript",
    revision: checkoutHead(repositoryRoot),
    model_id: application.embedder.modelId,
    repositories,
  });
  fs.mkdirSync(path.dirname(path.resolve(output)), { recursive: true });
  fs.writeFileSync(output, `${dumpJson(snapshot, { indent: 2 })}\n`, "utf8");
  process.stderr.write(`wrote ${output}\n`);
} finally {
  await application.store.close();
  if (dataDir === undefined) fs.rmSync(temporary, { recursive: true, force: true });
}

File: ts/packages/server/scripts/write_soak_snapshot.py
"""Write a Phase 9 soak snapshot with the Python build.

The migration plan's dual-run soak (§8) indexes the same real repositories
with both builds and diffs the chunk tables row by row plus compares search
rankings for a query set. This is the Python half: it produces the snapshot
artifact that ``write_soak_snapshot.ts`` mirrors and ``soak_compare.ts``
gates on. Run it *first* -- ``init_project`` writes the
``.ci-mcp/project.toml`` marker into each repository, and the project id in
that marker is what makes chunk ids comparable across the two builds.

Usage, from the repository root:

    uv run python ts/packages/server/scripts/write_soak_snapshot.py \
        --manifest soak.manifest.json --output soak-python.json

The manifest is the JSON shape ``src/soak.ts`` defines: repositories (path,
optional name) and queries. Indexing happens into a scratch data directory
removed afterwards (pass ``--data-dir`` to keep or reuse one), with execution
forced in-process and the daemon off, exactly like the benchmark commands.
"""

from __future__ import annotations

import argparse
import json
import shutil
import sys
import tempfile
from dataclasses import replace
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[4] / "src"))

from code_indexing_mcp import update_check
from code_indexing_mcp.application import Application, RuntimePaths
from code_indexing_mcp.settings import IndexSettings


def parse_arguments(argv: list[str]) -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--manifest", required=True, help="soak manifest JSON")
    parser.add_argument("--output", required=True, help="snapshot JSON to write")
    parser.add_argument(
        "--data-dir",
        default=None,
        help="reuse a specific scratch data directory instead of a temporary one",
    )
    return parser.parse_args(argv)


def main() -> None:
    arguments = parse_arguments(sys.argv[1:])
    manifest = json.loads(Path(arguments.manifest).read_text(encoding="utf-8"))
    if not manifest.get("queries"):
        raise SystemExit("the soak manifest needs at least one query")
    limit = int(manifest.get("limit", 8))
    if not 1 <= limit <= 50:
        raise SystemExit("the soak manifest limit must be from 1 to 50")

    temporary = (
        Path(arguments.data_dir)
        if arguments.data_dir is not None
        else Path(tempfile.mkdtemp(prefix="ci-mcp-py-soak-"))
    )
    settings = replace(
        IndexSettings.from_environment(),
        index_execution="in-process",
        broker_mode="off",
    )
    application = Application(
        RuntimePaths(data=temporary, cache=RuntimePaths.from_environment().cache),
        cwd=Path.cwd(),
        settings=settings,
    )
    try:
        repositories = []
        for repository in manifest["repositories"]:
            root = Path(repository["path"]).expanduser().resolve()
            name = repository.get("name") or root.name
            project = application.init_project(root)
            application.index_project(project.id)
            chunks = sorted(
                application.store.list_chunks([project.id]), key=lambda chunk: chunk.chunk_id
            )
            queries = []
            for query in manifest["queries"]:
                response = application.search_code(query, projects=[project.id], limit=limit)
                queries.append(
                    {
                        "query": query,
                        "hits": [
                            {"chunk_id": hit.chunk_id, "score": hit.score} for hit in response.hits
                        ],
                    }
                )
            print(f"{name}: {len(chunks)} chunks, {len(queries)} queries", file=sys.stderr)
            repositories.append(
                {
                    "name": name,
                    "path": str(root),
                    "project_id": project.id,
                    "chunk_count": len(chunks),
                    "chunks": [chunk.model_dump(mode="json") for chunk in chunks],
                    "queries": queries,
                }
            )
        snapshot = {
            "schema_version": 1,
            "build": "python",
            "revision": update_check.checkout_head(Path(__file__).resolve().parents[4]),
            "model_id": application.embedder.model_id,
            "repositories": repositories,
        }
        output = Path(arguments.output)
        output.parent.mkdir(parents=True, exist_ok=True)
        output.write_text(json.dumps(snapshot, indent=2) + "\n", encoding="utf-8")
        print(f"wrote {output}", file=sys.stderr)
    finally:
        if arguments.data_dir is None:
            shutil.rmtree(temporary, ignore_errors=True)


if __name__ == "__main__":
    main()
